### PR TITLE
Add learn-from-code-review skill

### LINKED ISSUE #86: "Learn from Code Review" Skill
This repo has an example of how you can learn from code reviews and distill the feedback back into a skill for the coding agent to do better next time: https://github.com/AmitPoonia/dstl-code-review

We could create a skill for this, such as "learn from code review", where the agent looks at all the code reviews on the repo over the past week or so, and then opens a draft PR with new skills for the repo (in the .agents/skills directory for specific skills, or in the repo AGENTS.md file for overall repo guidelines).

### PR DESCRIPTION
**Pull request by neubig**
## Summary

This PR adds a new skill that distills code review feedback from GitHub PRs into reusable skills and repository guidelines.

Closes #86

## What this skill does

The `learn-from-code-review` skill analyzes PR review comments from a repository and extracts recurring patterns of feedback. These patterns are transformed into:

- **Repository-specific skills** (`.openhands/skills/`) for domain-specific patterns (e.g., database queries, API design)
- **AGENTS.md updates** for general coding conventions and best practices

## Workflow

1. Fetch PR review comments from the target repository using GitHub API
2. Filter out noise (bot comments, low-signal responses like "LGTM", short comments)
3. Categorize comments by type (security, performance, style, architecture, etc.)
4. Identify recurring patterns (3+ similar comments)
5. Generate skills and/or AGENTS.md updates
6. Create a draft PR with the proposed changes

## Trigger phrases

- `/learn-from-reviews`
- "learn from code reviews"
- "distill reviews"

## Files changed

- `skills/learn-from-code-review/SKILL.md` - Main skill definition
- `skills/learn-from-code-review/README.md` - Documentation
- `marketplaces/default.json` - Added skill to marketplace

## Testing

- ✅ All existing tests pass (`test_skills_have_readme.py`)
- ✅ Marketplace JSON is valid

## Evidence

The skill was tested on the `OpenHands/software-agent-sdk` repository. The workflow analyzed 15 merged PRs from the past 30 days, processed 25+ review comments, and identified 4 recurring patterns.

**Test PR**: https://github.com/OpenHands/software-agent-sdk/pull/2280

The test generated:
- A new skill: `.openhands/skills/eval-risk-assessment/SKILL.md`
- Updates to `AGENTS.md` with a Code Review Learnings section

## Related

- Inspired by [dstl-code-review](https://github.com/AmitPoonia/dstl-code-review)
- Related skills: `code-review`, `github-pr-review`, `skill-creator`, `agent-memory`

### CODE REVIEW

File: skills/learn-from-code-review/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+- Code style/conventions
+- Architecture/design patterns
+- Error handling
+- Testing requirements
+- Documentation standards
+
+### Step 4: Distill Patterns
+
+For each category with sufficient examples (3+ similar comments), identify:
+
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/learn-from-code-review/SKILL.md:81]: This should be left to the agent's judgement, so this is fine.

File: skills/learn-from-code-review/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+gh api repos/{owner}/{repo}/pulls/{pr_number}/reviews \
+  --jq '.[] | select(.body != "") | {body: .body, user: .user.login, state: .state}'
+```
+
+### Step 3: Filter and Categorize Comments
+
+Apply noise filtering to keep only meaningful feedback:
+
+**Exclude:**
+- Bot comments (dependabot, copilot, github-actions, etc.)
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/learn-from-code-review/SKILL.md:64]: This comment is too pedantic, skip.

File: skills/learn-from-code-review/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+.openhands/skills/{domain-name}/SKILL.md
+```
+
+Example structure:
+```yaml
+---
+name: database-queries
+description: Database query patterns and best practices derived from code review feedback.
+---
+
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/learn-from-code-review/SKILL.md:97]: No, these are related actions so it doesn't matter.

File: skills/learn-from-code-review/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+| Include bots | No | Whether to include bot review comments |
+| Output format | both | Generate skills, AGENTS.md, or both |
+
+## Best Practices
+
+1. **Run periodically** - Schedule monthly or quarterly to capture evolving patterns
+2. **Review before merging** - Generated content is a draft; human review is essential
+3. **Iterate** - Refine patterns based on team feedback
+4. **Avoid duplication** - Check existing AGENTS.md and skills before adding
+5. **Cite sources** - Reference PR numbers when documenting patterns
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/learn-from-code-review/SKILL.md:165]: Let's fix this.

File: skills/learn-from-code-review/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+### Testing
+- Include edge case tests for user input validation
+- Mock external services in unit tests
+```
+
+### Step 6: Create Draft PR
+
+Open a draft PR with the proposed changes:
+
+```bash
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/learn-from-code-review/SKILL.md]: Yes, we can remove the git commands entirely.

File: skills/learn-from-code-review/README.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+- `GITHUB_TOKEN` environment variable
+- GitHub CLI (`gh`) available
+
+## Example
+
+```
+User: Learn from our code reviews over the past month
+
+Agent: I'll analyze recent PR review comments and distill them into actionable guidelines.
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/learn-from-code-review/README.md:30]: Yes, let's run a test where we have a sub-agent follow the workflow to give comments on OpenHands/software-agent-sdk and make sure that it works. The resulting PR should be should be cited in the "Evidence" section of the PR description in this PR.

File: skills/learn-from-code-review/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+---
+
+# Learn from Code Review
+
+Analyze code review comments from GitHub pull requests and distill them into reusable skills or repository guidelines that improve future code quality.
+
+## Overview
+
+Code review feedback contains valuable institutional knowledge that often gets buried across hundreds of PRs. This skill extracts meaningful patterns from review comments and transforms them into:
+
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/learn-from-code-review/SKILL.md:17]: Let's keep it as-is for now.

File: skills/learn-from-code-review/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+## Log Before Returning Errors
+
+```python
+logger.error(f"API error in {endpoint}: {e}", exc_info=True)
+return error_response(e)
+```
+```
+
+## Configuration Options
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/learn-from-code-review/SKILL.md]: I think this section can be deleted, and we can just say that it looks back over the past 30 days.

**Review by csmith49 — APPROVED** [review]
LGTM, curious to see what this will start producing.

Also interesting to see the skills referencing other skills.

File: skills/learn-from-code-review/SKILL.md [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+  --limit 50 \
+  --json number,title,mergedAt
+
+# For each PR, fetch review comments
+gh api repos/{owner}/{repo}/pulls/{pr_number}/comments \
+  --jq '.[] | {body: .body, path: .path, user: .user.login, created_at: .created_at}'
+
+# Also fetch review-level comments (not tied to specific lines)
+gh api repos/{owner}/{repo}/pulls/{pr_number}/reviews \
+  --jq '.[] | select(.body != "") | {body: .body, user: .user.login, state: .state}'
```
Comment (enyst) [inline skills/learn-from-code-review/SKILL.md:56]: @openhands The comment says reviews not tied to specific lines, but does that mean it doesn't get all inline review threads, or does it mean it gets them, but not the lines information where they were?
